In [5]:
import sys, io
import math
import numpy as np
import pandas as pd
from scipy.optimize import nnls
from fractions import Fraction
from sympy import factorint

In [6]:
def ChainPath(collatzNumber):
    path = []
    while collatzNumber != 1:
        if collatzNumber & 1 == 0:
            collatzNumber = collatzNumber // 2
            path.append("1")
        else:
            collatzNumber = (3 * collatzNumber + 1) // 2
            path.append("0")
    return "".join(path)
#
def fractionFromNodeTup(tup):
    p2, p3, c = tup
    fract = Fraction(2**p2 - c, 3**p3)
    return (fract.numerator, fract.denominator)
#

def TupChainFromPath(chain_path):
    tup_chain = [(0, 0, 0)]
    for chain_item in chain_path:
        p2, p3, c = tup_chain[-1]
        if chain_item == "1":
            tup_chain. append((p2 + 1, p3, c))
        else:
            tup_chain. append((p2 + 1, p3 + 1, c*3 + 2**p2))
            
        fract = fractionFromNodeTup(tup_chain[-1])
        print(fract)
    return tup_chain
#
def downUpTup(tup):
    p2, p3, c = tup
    p2_01, p3_01, c_01 = (p2 + 2, p3 + 1, c*3 + 2**p2)
    return (p2_01, p3_01, c_01)
#
def TupChain(collatzNumber):
    chain_path = ChainPath(collatzNumber)
    return TupChainFromPath(chain_path)
    return tup_chain
#

In [7]:
def ChainLength(collatzNumber):
    length = 0
    while collatzNumber != 1:
        if collatzNumber & 1 == 0:
            collatzNumber = collatzNumber // 2
            length += 1
        else:
            collatzNumber = (3 * collatzNumber + 1) // 2
            length += 1
    return length
#


In [8]:
# It very interesting how multiple powers of three originate in the same generation
for i in range(1, 70, 1):
    print((i, ChainLength(3**i)))

(1, 5)
(2, 13)
(3, 70)
(4, 16)
(5, 62)
(6, 24)
(7, 51)
(8, 51)
(9, 32)
(10, 89)
(11, 89)
(12, 89)
(13, 89)
(14, 97)
(15, 135)
(16, 89)
(17, 127)
(18, 97)
(19, 70)
(20, 70)
(21, 173)
(22, 276)
(23, 70)
(24, 173)
(25, 257)
(26, 219)
(27, 173)
(28, 173)
(29, 257)
(30, 257)
(31, 257)
(32, 303)
(33, 303)
(34, 303)
(35, 303)
(36, 303)
(37, 303)
(38, 303)
(39, 303)
(40, 303)
(41, 303)
(42, 333)
(43, 333)
(44, 303)
(45, 333)
(46, 333)
(47, 490)
(48, 333)
(49, 333)
(50, 333)
(51, 490)
(52, 333)
(53, 490)
(54, 333)
(55, 333)
(56, 333)
(57, 490)
(58, 490)
(59, 490)
(60, 490)
(61, 490)
(62, 485)
(63, 485)
(64, 485)
(65, 485)
(66, 528)
(67, 528)
(68, 528)
(69, 528)


In [15]:
def powerChainGroups(basis, offset, n):
    D = {}
    for i in range(1, n, 1):
        path = ChainPath((basis**i)+offset)
        l = len(path)
        if l not in D:
            D[l] = []
        D[l].append((i, path[::-1]))
    #
    return D
#

In [16]:
groups_3_0 = powerChainGroups(3, 0, 70)
groups_3_0

{5: [(1, '11100')],
 13: [(2, '1110110100010')],
 70: [(3,
   '1110111100011011101010111000011000000100100011010000100010010100000100'),
  (19,
   '1110110100110011101100001111111110111111101110010101111100110111000100'),
  (20,
   '1110110100110011101100001111111110111111101110010101111100110101011010'),
  (23,
   '1110110100110011101100001111111110111111101110010101110111111111110100')],
 16: [(4, '1110111100011010')],
 62: [(5, '11101111000110111010101110000110000001001000110100001000101100')],
 24: [(6, '111011010011011011110010')],
 51: [(7, '111011010011001110010011010110001011110000100110100'),
  (8, '111011010011001110010011010010110111010100101001010')],
 32: [(9, '11101101000111010101101111111100')],
 89: [(10,
   '11101101001101101101110100000001011101010011010100111100100101101001010101100110000000010'),
  (11,
   '11101101001101101101110100000001011101010011011100011000011110100001010001101110100100100'),
  (12,
   '111011010011011011011101000000010111010100110111000110000

# Collatz Chains of powers of 3

Collatz Chains of Powers of 3 often share the same chain length and terminate in the same series of steps, so there are certain starting paths in the lattice that lead to more powers of three than others:

## Shared starting paths of powers of three in the Collatz lattice by length

All Collatz lattice paths of powers of three with $length < 485$ start with '1110' and so are in the lower half of the upper sixteenth of the lattice.

<code>
{5: [(1, '11100')],
 13: [(2, '<span style="color: red;">111011</span>0100010')],
 16: [(4, '<span style="color: red;">111011</span>1100011010')],
 24: [(6, '<span style="color: red;">111011</span>010011011011110010')],
 32: [(9, '<span style="color: red;">111011</span>01000111010101101111111100')],
 51: [
  (7, '<span style="color: red;">111011010011001110010011010</span>110001011110000100110100'),
  (8, '<span style="color: red;">111011010011001110010011010</span>010110111010100101001010')],
 62: [(5, '11101111000110111010101110000110000001001000110100001000101100')],
 70: [
  ( 3, '<span style="color: red;">111011</span>1100011011101010111000011000000100100011010000100010010100000100'),
  (19, '<span style="color: red;">111011</span>0100110011101100001111111110111111101110010101111100110111000100'),
  (20, '<span style="color: red;">111011</span>0100110011101100001111111110111111101110010101111100110101011010'),
  (23, '<span style="color: red;">111011</span>0100110011101100001111111110111111101110010101110111111111110100')],
 89: [
  (10, '<span style="color: red;">1110110100110110110111010000000101110101001101</span>0100111100100101101001010101100110000000010'),
  (11, '<span style="color: red;">1110110100110110110111010000000101110101001101</span>1100011000011110100001010001101110100100100'),
  (12, '<span style="color: red;">1110110100110110110111010000000101110101001101</span>1100011000011110100001110000010110001111010'),
  (13, '<span style="color: red;">1110110100110110110111010000000101110101001101</span>1100011000001111011110010011100110010001100'),
  (16, '<span style="color: red;">1110110100110110110111010000000101110101001101</span>1100011000001111011110010010101101111101010')],
 97: [
  (14, '<span style="color: red;">1110110100011101010011101001101010111011111011100110000000111101010</span>110110101000001100101000010010'),
  (18, '<span style="color: red;">1110110100011101010011101001101010111011111011100110000000111101010</span>010101101010110011110101100010')],
 127: [(17, '1110110100111111011100110011000101110001110111010011000010110101010110010111000100110011001000000110110101111000010010000011100')],
 135: [(15, '111011010011001110010011010110011000111010111010000110111001111011000110110000100100110010000000001111000010011111000011010011010010100')],
 173: [
  (21, '<span style="color: red;">111011</span>11000110111010101110000110000001001000110100001000100101000001111101001100011000100110110100010101110010111110110000011110001110011000011111100011110001101001011101100'),
  (24,   '<span style="color: red;">111011</span>01001101101101110100000001110011101101101010001101100111010111000001001011011101000111101011011000110010111000110111101001100110101111111010000001011001000000000001010'),
  (27,   '<span style="color: red;">111011</span>01001101101101110100000001110011101101101010001101100111010111000001001011011101000111101011011000011111011010010001101101001001011100101000110010001001011111101100100'),
  (28,   '<span style="color: red;">111011</span>01001101101101110100000001110011101101101010001101100111010111000001001011011101000111101011011000011111011010010001101101001001011100101000110010001001011111100111010')],
 219: [(26,   '111011010011001110010011011111010111000001111000010001101011000100010000010000111011000101001000000101000000111011110010100110100010111111110100111011100000000011010010110011111110100101111001101001110111001110001000010')],
 257: [
  (25,   '<span style="color: red;">11101111000110111010101110000110000001001000110100001000100101000011000111111101011000001011100100111100011101000000001001100111111011111010101100101111000000101010011</span>010011110101011110011000010000110011000011100100001010111011000101000110011000000000111100'),
  (29,   '<span style="color: red;">11101111000110111010101110000110000001001000110100001000100101000011000111111101011000001011100100111100011101000000001001100111111011111010101100101111000000101010011</span>110000110100111001001111100010010100111100101101100000000110011110100010110100001111001100'),
  (30,   '<span style="color: red;">11101111000110111010101110000110000001001000110100001000100101000011000111111101011000001011100100111100011101000000001001100111111011111010101100101111000000101010011</span>110000110100111001001111100010010100111100101100100000011110101100110010100110110011010010'),
  (31,   '<span style="color: red;">11101111000110111010101110000110000001001000110100001000100101000011000111111101011000001011100100111100011101000000001001100111111011111010101100101111000000101010011</span>110000110100111001001111100010010100111100101100100000011110101100110010100110011111010100')],
 276: [  (22,   '<span style="color: red;">111011</span>110001101110101011100001100000010010001101000010001001110000011010100010100111001111100101100100001011011010001100011110000001001001010011111001000010111001111110010101100000000100011100101100000111001000100100100101110100101000111000000100011011011101001101010100110010')],
 303: [
  (32,   '<span style="color: red;">111011</span>010011001110010011011111010111000001111000010001101011000100010000010001111110100101001001011010101000010011000101110111110010010011010010011101010110111110001001010110011100010000000000110011110110000000000111011110110011010100110010110001001001010101011111100110110100000100010101011100000101010'),
  (33,   '<span style="color: red;">111011</span>010011001110010011011111010111000001111000010001101011000100010000010001111110100101001001011010101000010011000101110111110010010011010010011101010110111110001001010110011100010000000000110011110110000000000111011110110011010100110010110001001001010101011111100110110100000100010101010100011011100'),
  (34,   '<span style="color: red;">111011</span>010011001110010011011111010111000001111000010001101011000100010000010001111110100101001001011010101000010011000101110111110010010011010010011101110001011011000111100100010010110000000001111011001111100001111000101001110110110101110101110011001001100010111100011000111001100000001001001011000100010'),
  (35,   '<span style="color: red;">111011</span>010011001110010011011111010111000001111000010001101011000100010000010001111110100101001001011010101000010011000101110111110010010011010010011101110001001111101100001011110001000010010010010000100010011111101100100110101111001001111100001001111100011111010001001100100100100111111000100001110000100'),
  (36,   '<span style="color: red;">111011</span>010011001110010011011111010111000001111000010001101011000100010000010001111110100101001001011010101000010011000101110111110010010011010010011101110001001111101100001011110001000010010010010000100010011111101100100110101111001001111100001001111100011111010001001100100100100111111000100001010011010'),
  (37,   '<span style="color: red;">111011</span>010011001110010011011111010111000001111000010001101011000100010000010001111110100101001001011010101000010011000101110111110010010011010010011101110001001111101100001011110001000010010010010000100010011111101100100110101111001001111100001001111100011111010001001100100101100000111011101001010101100'),
  (38,   '<span style="color: red;">111011</span>0100110011100100110111110101110000011110000100011010110001000100000100011111101001010010010110101010000100110001011101111100100100110100100111011</span>10001001111101100001011110001000010010010010000100010011111101100100110101111001001111100001001111100011111010001001100100101100001100100111110101110010'),
  (39,   '<span style="color: red;">111011</span>010011001110010011011111010111000001111000010001101011000100010000010001111110100101001001011010101000010011000101110111110010010011010010011101110001011001000111110000110111111000001100111010110010100010101101011110000010001010110101001110011011011101110110000100010000101001111111000001110110100'),
  (40,   '<span style="color: red;">111011</span>010011001110010011011111010111000001111000010001101011000100010000010001111110100101001001011010101000010011000101110111110010010011010010011101110001001111101100001011110001000010010010010000100010011111101100100110101111001001111100001001111100110001011100111110110101011010001101010110011001010'),
  (41,   '<span style="color: red;">111011</span>010011001110010011011111010111000001111000010001101011000100010000010001111110100101001001011010101000010011000101110111110010010011010010011101110001001111101100001011110001000010010010010000100010011111101100100110101111001001111100001001111100110001011100111110110101011010001101010110001111100'),
  (44,   '<span style="color: red;">111011</span>010011001110010011011111010111000001111000010001101011000100010000010001111110100101001001011010101000010011000101110111110010010011010010011101110001001111101100001011110001000010010010010000100010011111101100100110101111001001111100001001111100110001011101100101110100101110111101011100111111010')],
 333: [(42,   '<span style="color: red;">111011</span>010111111111101001010101101011111000100001001001011111001110101110010111000110110001001110111101001000010110111001111100000001001001010101101101010000100101010110001001111110010010001110110001010000000000110010100101101001100000100111111001100111010110001001100001011010101011001010000101010110001010111001011110100110110000010'),
  (43,   '<span style="color: red;">111011</span>010111111111101001010101101011111000100001001001011111001110101110010111000110110001001110111101001000110001010110011100001100111110111011110000011010010000000010101111110011110000110110101000001001000001101100111010101001110110001101011101110011010100001110000000001100010001100111100010110011010111010000001100001100010100100'),
  (45,   '<span style="color: red;">111011</span>010111111111101001010101101011111000100001001001011111001110101110010111000110110001001110111101001000010110111001111100000001001001010101101101010000100101010110001001111110010010001110110001010000000000110010100101101001100000100111111000111011111010011100101010100010110111101001011101110010101100000010010011011101000001100'),
  (46,   '<span style="color: red;">111011</span>010111111111101001010101101011111000100001001001011111001110101110010111000110110001001110111101001000110001010110011100001100111110111011110000011010010000000010101111110011110000110110101000001001000001101100111010101001110110001101011101110011010100001110000000001100010001100111100010010110111101011000010100101011010010010'),
  (48,   '<span style="color: red;">111011</span>010111111111101001010101101011111000100001001001011111001110101110010111000110110001001110111101001000010110111001111100000001001001010101101101010000100101010110001001111110010010001110110001010000000000110010100101101001100000100111111000111110001111101110111101110101100011000111101010010011111000101001000000101110001101010'),
  (49,   '<span style="color: red;">111011</span>010111111111101001010101101011111000100001001001011111001110101110010111000110110001001110111101001000010110111001111100000001001001010101101101010000100101010110001001111110010010001110110001010000000000110010100101101001100000100111111000111110011001111000101010110100100100011110010111111011011001100111101100110100010011100'),
  (50,   '<span style="color: red;">111011</span>010111111111101001010101101011111000100001001001011111001110101110010111000110110001001110111101001000010110111001111100000001001001010101101101010000100101010110001001111110010010001110110001010000000000110010100101101001100000100111111000111110011001111000101010110100100100011110010111111011001111111001111000100110011100010'),
  (52,   '<span style="color: red;">111011</span>010111111111101001010101101011111000100001001001011111001110101110010111000110110001001110111101001000010110111001111100000001001001010101101101010000100101010110001001111110010010001110110001010000000000110010100101101001100000100111111000111110011001111000101010110100100100011110010111111011001111111001111000100011111011010'),
  (54,   '<span style="color: red;">111011</span>010111111111101001010101101011111000100001001001011111001110101110010111000110110001001110111101001000010110111001111100000001001001010101101101010000100101010110001001111110010010001110110001010000000000110010100101101001100000100111111000111110011001111000101010110100100100011110010111111011001111101111101011111110010110010'),
  (55,   '<span style="color: red;">111011</span>010111111111101001010101101011111000100001001001011111001110101110010111000110110001001110111101001000010110111001111100000001001001010101101101010000100101010110001001111110010010001110110001010000000000110010100101101001100000100111111000111110011001111000101010110100100100011110010111111011001111101111101011111011101110100'),
  (56,   '<span style="color: red;">111011</span>010111111111101001010101101011111000100001001001011111001110101110010111000110110001001110111101001000110001010110011100001100111110111011110000011010010000000010101111110011110000110110101000001001000001101100111010101001110010111111011010110001001100010000010010011111011111111101010000101110100010110011111101111110010001010')],
 485: [
  (62,   '<span style="color: red;">111111111011111101111001001010101101010110010100000100110111001010100111111000110100100001001011000001100010001111010100100011111010111011000000000100011111101001110101100100000011011111110001011010101001100100111100000000101111101101000101001000010100010111010100101111101000110001001101011011100001101011100110110110110101000100000110010101110100101001000001110011011101110110</span>11000101100111101101011001001000010101001000110011001110101000110111000001101010110010000100111101001010010'),
  (63,
   '<span style="color: red;">111111111011111101111001001010101101010110010100000100110111001010100111111000110100100001001011000001100010001111010100100011111010111011000000000100011111101001110101100100000011011111110001011010101001100100111100000000101111101101000101001000010100010111010100101111101000110001001101011011100001101011100110110110110101000100000110010101110100101001000001110011011101110110</span>01011011101011011011111011111001010011100111100000001000010001100000001011000011011110101010011000001010100'),
  (64,
   '<span style="color: red;">111111111011111101111001001010101101010110010100000100110111001010100111111000110100100001001011000001100010001111010100100011111010111011000000000100011111101001110101100100000011011111110001011010101001100100111100000000101111101101000101001000010100010111010100101111101000110001001101011011100001101011100110110110110101000100000110010101110100101001000001110011011101110110</span>01011011101011011011111011111001010011100111100000001000010001100000001011000011011110101010001000110101010'),
  (65,
   '<span style="color: red;">111111111011111101111001001010101101010110010100000100110111001010100111111000110100100001001011000001100010001111010100100011111010111011000000000100011111101001110101100100000011011111110001011010101001100100111100000000101111101101000101001000010100010111010100101111101000110001001101011011100001101011100110110110110101000100000110010101110100101001000001110011011101110110</span>11000101100111101101011001001000010101001000110001111100111111000101100111101110000100000111100001001011100')],
 490: [
  (47,   '<span style="color: red;">1110111100011011101010111000011000000100100011010000100010010100000101111011010110101001100011100010011000110001111010011111110011101101110001000100010100001010100100101101110000000010001000110101011110111001101010101101101011001101001101011000000111000111</span>011111000001111111100010001011001110000101011010010010001000011011111100101000110101000010011100010111001010101011000010100110101100110001010100000100001111000000111010010001011101001000001111000001100111010110110001001011010000010100'),
  (51,   '<span style="color: red;">1110111100011011101010111000011000000100100011010000100010010100000101111011010110101001100011100010011000110001111010011111110011101101110001000100010100001010100100101101110000000010001000110101011110111001101010101101101011001101001101011000000111000111</span>110110010101010001010100010011111001101000100101000001010100010001011111000011101100110111101011111011110011110111000000010110110010010101000110001011011000001110011010000101100100000011000111100111100001100110000011010000110101000100'),
  (53,   '<span style="color: red;">1110111100011011101010111000011000000100100011010000100010010100000101111011010110101001100011100010011000110001111010011111110011101101110001000100010100001010100100101101110000000010001000110101011110111001101010101101101011001101001101011000000111000111</span>110110010101010001010100010011111001101000100101000001010100010001011111000011101100110111101011111011110011110111000000010110110010010101000110001011011000001110011010000101100100000011000111100111100001100010000110110101100001101100'),
  (57,   '<span style="color: red;">1110111100011011101010111000011000000100100011010000100010010100000101111011010110101001100011100010011000110001111010011111110011101101110001000100010100001010100100101101110000000010001000110101011110111001101010101101101011001101001101011000000111000111</span>110110010101010001010100010011111001101000100101000001010100010001011111000011101100110111101011111011110011110111000000010011101011000011010010100110101011010111010010010110011101110111011111000010001001000010100000000100100110111100'),
  (58,   '<span style="color: red;">1110111100011011101010111000011000000100100011010000100010010100000101111011010110101001100011100010011000110001111010011111110011101101110001000100010100001010100100101101110000000010001000110101011110111001101010101101101011001101001101011000000111000111</span>110110010101010001010100010011111001101000100101000001010100010001011111000011101100110111101011111011110011110111000000010011101011000011010010100110101011010111010010010110011111001000011101101111000111100000101010010001010011000010'),
  (59,   '<span style="color: red;">1110111100011011101010111000011000000100100011010000100010010100000101111011010110101001100011100010011000110001111010011111110011101101110001000100010100001010100100101101110000000010001000110101011110111001101010101101101011001101001101011000000111000111</span>110110010101010001010100010011111001101000100101000001010100010001011111000011101100110111101011111011110011110111000000010011101011000011010010100110101011011101100100100100110101000001100111010111001100100100110010010011110111100100'),
  (60,   '<span style="color: red;">1110111100011011101010111000011000000100100011010000100010010100000101111011010110101001100011100010011000110001111010011111110011101101110001000100010100001010100100101101110000000010001000110101011110111001101010101101101011001101001101011000000111000111</span>110110010101010001010100010011111001101000100101000001010100010001011111000011101100110111101011111011110011110111000000010011101011000011010010100110101011011101100100100100110101000001100111010111001100100100110010010011110110111010'),
  (61,   '<span style="color: red;">1110111100011011101010111000011000000100100011010000100010010100000101111011010110101001100011100010011000110001111010011111110011101101110001000100010100001010100100101101110000000010001000110101011110111001101010101101101011001101001101011000000111000111</span>110110010101010001010100010011111001101000100101000001010100010001011111000011101100110111101011111011110011110111000000010011101011000011010010100110101011010111010010010110011111001000011101101101000111111101011011000100100001001100')],
 528: [
  (66,   '<span style="color: red;">11101111000110111010101110000110000001001000110100001000100101000111101110110101110101101011110101110100010101010010111100000111111100100000011101110101011000010000101100100000110010100101100011111111001011001010011110011001000001011101110010010110101110000110001000</span>0111111111110001110101100111100011011011011101011001100010110101011010101001100110100010101000000101001101011000100100001010111100101111111111101000000001100010000110101001110000001000110110011101110101010010001101100011101011111100100010101100100000001010100010'),
  (67,   '<span style="color: red;">11101111000110111010101110000110000001001000110100001000100101000111101110110101110101101011110101110100010101010010111100000111111100100000011101110101011000010000101100100000110010100101100011111111001011001010011110011001000001011101110010010110101110000110001000</span>1100101110111100011010100100111010101110100101110001111100011001011111001000111000000100100111111110010100001100001100110111111111011010101000101000000110011100010011111011000100101110011001001100011111011001000011010110010001001000101001110111011001100000000100'),
  (68,   '<span style="color: red;">11101111000110111010101110000110000001001000110100001000100101000111101110110101110101101011110101110100010101010010111100000111111100100000011101110101011000010000101100100000110010100101100011111111001011001010011110011001000001011101110010010110101110000110001000</span>1100101110111100011010100100111010101110100101110001111100011001011111001000111000000100100111111110010100001100001100110111111111011010101000101000000110011100010011111011000100101110011001001100011111011001000011010110010001001000101001110111011000100000011010'),
  (69,   '<span style="color: red;">11101111000110111010101110000110000001001000110100001000100101000111101110110101110101101011110101110100010101010010111100000111111100100000011101110101011000010000101100100000110010100101100011111111001011001010011110011001000001011101110010010110101110000110001000</span>0111111111110001110101100111100011011011011101011001100010110101011010101001100110100010101000000101001101011000100100001010111100101111111111101000000001100010000110101001110000001000110110110010110101000110111111001100011100101100101110001010100111000100101100')]
}

</code>

In [18]:
# It very interesting how multiple powers of three originate in the same generation
for i in range(1, 70, 1):
    print((i, ChainLength((3**i)-1)))

(1, 1)
(2, 3)
(3, 8)
(4, 8)
(5, 62)
(6, 62)
(7, 24)
(8, 24)
(9, 32)
(10, 32)
(11, 89)
(12, 89)
(13, 89)
(14, 89)
(15, 70)
(16, 70)
(17, 127)
(18, 127)
(19, 97)
(20, 97)
(21, 173)
(22, 173)
(23, 276)
(24, 276)
(25, 257)
(26, 257)
(27, 219)
(28, 219)
(29, 257)
(30, 257)
(31, 257)
(32, 257)
(33, 303)
(34, 303)
(35, 303)
(36, 303)
(37, 303)
(38, 303)
(39, 303)
(40, 303)
(41, 303)
(42, 303)
(43, 333)
(44, 333)
(45, 333)
(46, 333)
(47, 333)
(48, 333)
(49, 333)
(50, 333)
(51, 333)
(52, 333)
(53, 490)
(54, 490)
(55, 333)
(56, 333)
(57, 490)
(58, 490)
(59, 490)
(60, 490)
(61, 490)
(62, 490)
(63, 490)
(64, 490)
(65, 485)
(66, 485)
(67, 406)
(68, 406)
(69, 528)


In [12]:
# It very interesting how multiple powers of three originate in the same generation
for i in range(1, 70, 1):
    print((i, ChainLength((3**i)-2)))

(1, 0)
(2, 11)
(3, 16)
(4, 24)
(5, 16)
(6, 32)
(7, 70)
(8, 35)
(9, 89)
(10, 35)
(11, 70)
(12, 135)
(13, 70)
(14, 89)
(15, 173)
(16, 97)
(17, 173)
(18, 219)
(19, 173)
(20, 70)
(21, 173)
(22, 70)
(23, 257)
(24, 219)
(25, 219)
(26, 219)
(27, 173)
(28, 219)
(29, 257)
(30, 219)
(31, 341)
(32, 303)
(33, 303)
(34, 303)
(35, 303)
(36, 303)
(37, 303)
(38, 303)
(39, 303)
(40, 333)
(41, 303)
(42, 333)
(43, 333)
(44, 490)
(45, 333)
(46, 333)
(47, 490)
(48, 333)
(49, 333)
(50, 333)
(51, 333)
(52, 490)
(53, 490)
(54, 490)
(55, 490)
(56, 406)
(57, 490)
(58, 406)
(59, 490)
(60, 485)
(61, 490)
(62, 406)
(63, 406)
(64, 490)
(65, 406)
(66, 528)
(67, 528)
(68, 490)
(69, 528)


In [13]:
for i in range(1, 20, 1):
    print((i, ChainLength(((2**i)-1)*(3**i))))

(1, 5)
(2, 70)
(3, 68)
(4, 104)
(5, 59)
(6, 114)
(7, 115)
(8, 78)
(9, 82)
(10, 80)
(11, 81)
(12, 139)
(13, 140)
(14, 163)
(15, 115)
(16, 292)
(17, 166)
(18, 294)
(19, 268)


In [14]:
# My longest-chain approach
x = 27
for i in range(1, 10, 1):
    l = ChainLength(x)
    print((i, l, x))
    x = (2**l) + x

(1, 70, 27)
(2, 373, 1180591620717411303451)
(3, 1527, 19239260838083241802870625048898248928261591440656956380834127638791856333738872368854622195948616835955022626843)
(4, 6897, 4707641458830141774570540246149051001781432124893923733124365310674330842952740626306806687709660253248923371642359384351293783161708902978456275366974169241123463553479736767130433745932971679056338951440210586049994816128719920135666771722850601494136358674496681010995065147034409039866785476499581778067126892791448421763777796897583512597813651246782129086591201861790374349775324870585333967859233385391296678377243838506208392365167106794291042445164571)
(5, 32637, 15991164633320783658632168629237223905249809332997960747134935388853238699045898334376347616982564111114794311678413638258527613815754164777108193613436145852107367760652502441347061754233843581065478141045690744885060834319241957893152578659473843728008164976306146368161285208237679898047848480607031379324054738890161373755550264834622841662614

ValueError: Exceeds the limit (4300 digits) for integer string conversion; use sys.set_int_max_str_digits() to increase the limit